File: src/core_ffi/rust_ffi/adapter/layout.cc
#include "rust_ffi/include/layout.h"

#include <group/group.h>
#include <mover/mover.h>
#include <common_code/common_code.h>

using klotski::codec::RawCode;
using klotski::codec::CommonCode;

using klotski::mover::MaskMover;

using klotski::group::Group;
using klotski::group::BLOCK_NUM;
using klotski::group::GroupCases;
using klotski::group::GroupUnion;

using klotski::ffi::RsLayout;
using klotski::ffi::RsShortCode;

// TODO: it seems that cxx.rs not support `std::optional`
uint64_t klotski::ffi::layout_from_str(const rust::Str s) {
    const std::string_view sv {s.data(), s.length()};
    if (const auto ret = CommonCode::from_string(sv); ret.has_value()) {
        return ret.value().unwrap();
    }
    return 0x10FFFFFFFF; // return invalid value for now
}

bool klotski::ffi::layout_check(const uint64_t val) {
    return CommonCode::check(val);
}

rust::String klotski::ffi::layout_to_str(uint64_t val) {
    const auto layout = RsLayout {val};
    return layout.to_string();
}

rust::String klotski::ffi::layout_to_shorten_str(uint64_t val) {
    const auto layout = RsLayout {val};
    return layout.to_shorten_string();
}

uint32_t klotski::ffi::layout_to_short_code(uint64_t val) {
    const auto layout = RsLayout {val};
    return layout.to_short_code().code;
}

bool klotski::ffi::layout_is_horizontal_mirror(uint64_t val) {
    const auto layout = RsLayout {val};
    return layout.is_horizontal_mirror();
}

bool klotski::ffi::layout_is_vertical_mirror(uint64_t val) {
    const auto layout = RsLayout {val};
    return layout.is_vertical_mirror();
}

uint64_t klotski::ffi::layout_to_horizontal_mirror(uint64_t val) {
    const auto layout = RsLayout {val};
    return layout.to_horizontal_mirror().code;
}

uint64_t klotski::ffi::layout_to_vertical_mirror(uint64_t val) {
    const auto layout = RsLayout {val};
    return layout.to_vertical_mirror().code;
}

uint8_t klotski::ffi::layout_n_1x1(uint64_t val) {
    const auto layout = RsLayout {val};
    return layout.n_1x1();
}

uint8_t klotski::ffi::layout_n_1x2(uint64_t val) {
    const auto layout = RsLayout {val};
    return layout.n_1x2();
}

uint8_t klotski::ffi::layout_n_2x1(uint64_t val) {
    const auto layout = RsLayout {val};
    return layout.n_2x1();
}

uint8_t klotski::ffi::layout_type_id(uint64_t val) {
    const auto layout = RsLayout {val};
    return layout.type_id();
}

uint16_t klotski::ffi::layout_pattern_id(uint64_t val) {
    const auto layout = RsLayout {val};
    return layout.pattern_id();
}

uint8_t klotski::ffi::layout_toward_char(uint64_t val) {
    const auto layout = RsLayout {val};
    return layout.toward_char();
}

uint32_t klotski::ffi::layout_case_id(uint64_t val) {
    const auto layout = RsLayout {val};
    return layout.case_id();
}

rust::Vec<uint64_t> klotski::ffi::layout_next_cases(uint64_t val) {
    std::vector<CommonCode> result;
    auto mover = MaskMover([&result](const RawCode code, uint64_t) {
        result.emplace_back(code.to_common_code());
    });
    mover.next_cases(CommonCode::unsafe_create(val).to_raw_code(), 0);

    rust::Vec<uint64_t> vec;
    for (auto x : result) {
        vec.emplace_back(x.unwrap());
    }
    return vec;
}

rust::String RsLayout::to_string() const noexcept {
    return CommonCode::unsafe_create(code).to_string();
}

rust::String RsLayout::to_shorten_string() const noexcept {
    return CommonCode::unsafe_create(code).to_string(true);
}

RsShortCode RsLayout::to_short_code() const noexcept {
    return {CommonCode::unsafe_create(code).to_short_code().unwrap()};
}

bool RsLayout::is_vertical_mirror() const noexcept {
    return CommonCode::unsafe_create(code).is_vertical_mirror();
}

bool RsLayout::is_horizontal_mirror() const noexcept {
    return CommonCode::unsafe_create(code).is_horizontal_mirror();
}

RsLayout RsLayout::to_vertical_mirror() const noexcept {
    return {CommonCode::unsafe_create(code).to_vertical_mirror().unwrap()};
}

RsLayout RsLayout::to_horizontal_mirror() const noexcept {
    return {CommonCode::unsafe_create(code).to_horizontal_mirror().unwrap()};
}

uint8_t RsLayout::n_1x1() const noexcept {
    return std::get<2>(BLOCK_NUM[type_id()]);
}

uint8_t RsLayout::n_1x2() const noexcept {
    return std::get<0>(BLOCK_NUM[type_id()]) - n_2x1();
}

uint8_t RsLayout::n_2x1() const noexcept {
    return std::get<1>(BLOCK_NUM[type_id()]);
}

uint8_t RsLayout::n_2x2() const noexcept {
    return 1;
}

uint8_t RsLayout::type_id() const noexcept {
    return GroupUnion::from_common_code(CommonCode::unsafe_create(code)).unwrap();
}

uint16_t RsLayout::pattern_id() const noexcept {
    return GroupCases::obtain_group(CommonCode::unsafe_create(code)).pattern_id();
}

uint8_t RsLayout::toward_char() const noexcept {
    return GroupCases::obtain_group(CommonCode::unsafe_create(code)).toward_char();
}

uint32_t RsLayout::case_id() const noexcept {
    return GroupCases::obtain_info(CommonCode::unsafe_create(code)).case_id();
}

rust::Vec<RsLayout> RsLayout::next_cases() const noexcept {
    std::vector<CommonCode> result;
    auto mover = MaskMover([&result](const RawCode code, uint64_t) {
        result.emplace_back(code.to_common_code());
    });
    mover.next_cases(CommonCode::unsafe_create(code).to_raw_code(), 0);

    rust::Vec<RsLayout> vec;
    for (auto x : result) {
        vec.emplace_back(RsLayout(x.unwrap()));
    }
    return vec;
}

File: src/core_ffi/rust_ffi/src/layout.rs
use crate::ffi;
use crate::ShortCode;

/// Layout represents a valid klotski situation, which fills single `2x2` block
/// and any number of `1x1` / `1x2` / `2x1` blocks without overlapping within
/// the range of `5x4`, and ensures that at least two spaces need to remain.
/// These layout information will be stored in a 36-bit value (u64), which can
/// uniquely mark a valid klotski situation, and any valid situation can also
/// be represented by a unique value.
#[derive(Debug)]
pub struct Layout {
    value: u64
}

impl Layout {
    /// Check whether the klotski layout value is a valid.
    pub fn check(val: u64) -> bool {
        ffi::layout_check(val)
    }

    /// Create klotski layout without any check. Note that if an invalid value
    /// is passed, unpredictable behavior may occur.
    pub fn unsafe_new(val: u64) -> Self {
        Self { value: val }
    }

    /// Create klotski layout with validity check.
    pub fn new(val: u64) -> Option<Self> {
        if !Self::check(val) {
            return None
        }
        Some(Self::unsafe_new(val))
    }

    /// Create klotski layout from string form.
    pub fn from_str(msg: &str) -> Option<Self> {
        let val = ffi::layout_from_str(msg);
        if val == 0x10_FFFF_FFFF {
            return None;
        }
        Some(Self::unsafe_new(val))
    }
}

impl Layout {
    /// Get the original value of klotski layout.
    pub fn unwrap(&self) -> u64 {
        self.value
    }

    pub fn to_string(&self) -> String {
        ffi::layout_to_str(self.value)
    }

    pub fn to_shorten_string(&self) -> String {
        ffi::layout_to_shorten_str(self.value)
    }

    pub fn to_short_code(&self) -> ShortCode {
        ShortCode::unsafe_new(ffi::layout_to_short_code(self.value))
    }
}

impl Into<u64> for Layout {
    fn into(self) -> u64 {
        self.unwrap()
    }
}

impl Into<String> for Layout {
    fn into(self) -> String {
        self.to_string()
    }
}

impl Into<ShortCode> for Layout {
    fn into(self) -> ShortCode {
        self.to_short_code()
    }
}

impl Layout {
    pub fn is_horizontal_mirror(&self) -> bool {
        ffi::layout_is_horizontal_mirror(self.value)
    }

    pub fn is_vertical_mirror(&self) -> bool {
        ffi::layout_is_vertical_mirror(self.value)
    }

    pub fn to_vertical_mirror(&self) -> Layout {
        let val = ffi::layout_to_vertical_mirror(self.value);
        Layout::unsafe_new(val)
    }

    pub fn to_horizontal_mirror(&self) -> Layout {
        let val = ffi::layout_to_horizontal_mirror(self.value);
        Layout::unsafe_new(val)
    }
}

impl Layout {
    pub fn n_1x1(&self) -> u8 {
        ffi::layout_n_1x1(self.value)
    }

    pub fn n_1x2(&self) -> u8 {
        ffi::layout_n_1x2(self.value)
    }

    pub fn n_2x1(&self) -> u8 {
        ffi::layout_n_2x1(self.value)
    }

    pub fn n_2x2(&self) -> u8 {
        1
    }
}

impl Layout {
    pub fn type_id(&self) -> u8 {
        ffi::layout_type_id(self.value)
    }

    pub fn pattern_id(&self) -> u16 {
        ffi::layout_pattern_id(self.value)
    }

    pub fn toward_char(&self) -> u8 {
        ffi::layout_toward_char(self.value)
    }

    pub fn case_id(&self) -> u32 {
        ffi::layout_case_id(self.value)
    }
}

impl Layout {
    pub fn next_cases(&self) -> Vec<Layout> {
        ffi::layout_next_cases(self.value)
            .iter()
            .map(|val| Layout::unsafe_new(*val))
            .collect()
    }
}

impl PartialEq for Layout {
    fn eq(&self, other: &Self) -> bool {
        self.value == other.value
    }
}

// TODO: add `Display` trait support
